lib/src/command/upgrade.dart

```diff
@@ -140,9 +167,61 @@ Consider using the Dart 2.19 sdk to migrate to null safety.''');
       summaryOnly: onlySummary,
       analytics: analytics,
     );
+
     _showOfflineWarning();
   }
 
+  /// Updates [changes] to contain changes that updates constraints in [pubspec]
+  /// to have their lower bound match the version in [packages].
+  ///
+  /// If packages to update where given on the commandline, only those are
+  /// tightened. Otherwise all packages are tightened.
+  ///
+  /// If dependency has already been updated in [changes], the update will apply
+  /// on top of that change.
+  void tighten(
+    Pubspec pubspec,
+    List<PackageId> packages,
+    Map<PackageRange, PackageRange> changes,
+  ) {
+    if (argResults.flag('example') && entrypoint.example != null) {
+      log.warning(
+        'Running `upgrade --tighten` only in `${entrypoint.rootDir}`. Run `$topLevelProgram pub upgrade --tighten --directory example/` separately.',
+      );
+    }
+    final toTighten = argResults.rest.isEmpty
```

Comment:
nit: Maybe we should have a getter:
```dart
/// List of package names to upgrade, if empty then upgrade all packages.
///
/// This allows the user to specify list of names that they want the
/// upgrade command to affect.
List<String> get packagesToUpgrade => argResults.rest;
```

We could also call it `targetPackages`, just a suggestion.